packages/analytics/src/get-config.ts

```diff
@@ -0,0 +1,319 @@
+  };
+
+  const signal = new AnalyticsAbortSignal();
+
+  setTimeout(
+    async () => {
+      // Note a very low delay, eg < 10ms, can elapse before listeners are initialized.
+      signal.abort();
+    },
+    timeoutMillis !== undefined ? timeoutMillis : FETCH_TIMEOUT_MILLIS
+  );
+
+  return attemptFetchDynamicConfigWithRetry(
+    { appId, apiKey, measurementId },
+    throttleMetadata,
+    signal,
+    retryData
+  );
+}
+
+/**
+ * Runs one retry attempt.
+ * @param appFields Necessary app config fields.
+ * @param throttleMetadata Ongoing metadata to determine throttling times.
+ * @param signal Abort signal.
+ */
+async function attemptFetchDynamicConfigWithRetry(
+  appFields: AppFields,
+  { throttleEndTimeMillis, backoffCount }: ThrottleMetadata,
+  signal: AnalyticsAbortSignal,
+  retryData: RetryData = defaultRetryData // for testing
+): Promise<DynamicConfig | MinimalDynamicConfig> {
+  const { appId, measurementId } = appFields;
+  // Starts with a (potentially zero) timeout to support resumption from stored state.
+  // Ensures the throttle end time is honored if the last attempt timed out.
+  // Note the SDK will never make a request if the fetch timeout expires at this point.
+  try {
+    await setAbortableTimeout(signal, throttleEndTimeMillis);
+  } catch (e) {
+    if (measurementId) {
+      logger.warn(
+        `Timed out fetching this Firebase project's measurement id from the server.` +
+          ` Falling back to measurement id ${measurementId}` +
+          ` provided in "measurementId" field. [${e.message}]`
+      );
+      return { appId, measurementId };
+    }
+    throw e;
+  }
+
+  try {
+    const response = await fetchDynamicConfig(appFields);
+
+    // Note the SDK only clears throttle state if response is success or non-retriable.
+    retryData.deleteThrottleMetadata(appId);
+
+    return response;
+  } catch (e) {
+    if (!isRetriableError(e)) {
+      retryData.deleteThrottleMetadata(appId);
+      if (measurementId) {
+        logger.warn(
+          `Failed to fetch this Firebase project's measurement id from the server.` +
+            ` Falling back to measurement id ${measurementId}` +
+            ` provided in "measurementId" field. [${e.message}]`
+        );
+        return { appId, measurementId };
+      } else {
+        throw e;
+      }
+    }
+
+    const backoffMillis =
+      Number(e.httpStatus) === 503
+        ? calculateBackoffMillis(
+            backoffCount,
+            retryData.intervalMillis,
+            LONG_RETRY_FACTOR
+          )
+        : calculateBackoffMillis(backoffCount, retryData.intervalMillis);
+
+    // Increments backoff state.
+    const throttleMetadata = {
+      throttleEndTimeMillis: Date.now() + backoffMillis,
+      backoffCount: backoffCount + 1
+    };
+
+    // Persists state.
+    retryData.setThrottleMetadata(appId, throttleMetadata);
+    logger.debug(`Calling attemptFetch again in ${backoffMillis} millis`);
```

Comment:
I didn't know that, I thought debug was something we could ask a user to turn on if we wanted to help them debug something. Or is that `info`?

Reply:
Hmm yes, but I think we ought to stop doing that in the prod build and provide a special debug build for debugging.   Otherwise the debug messages are mostly dead code. It's probably worth discussing with other teams before changing it. So let's keep it for now.